CONTRIBUTING.md

```diff
@@ -14,19 +17,67 @@ Conduct](https://www.rust-lang.org/policies/code-of-conduct).
+getters and that don't break the abstraction can be public, since this crate does not provide a library API. Public-ness is used mostly as a marker that something is an implementation detail of a module.
+
+Please run `cargo fmt` and `cargo clippy`. These are checked in CI.
 
 ## Testing
 
-Of course, please add tests for new features or bug fixes. The primary means of testing is Rust source trees under `testdata/tree`: you can copy an existing tree and modify it to show the new behavior that you want to test.
+Of course, please add tests for new features or bug fixes.
+
+Cargo-mutants is primarily tested on its public interface, which is the command line. These tests live in `tests/cli` and generally have the form of:
+
+1. Make a copy of a `testdata` tree, so that it's not accidentally modified.
+2. Run a `cargo-mutants` command on it.
+3. Inspect the stdout, return code, or `mutants.out`.
+
+`cargo-mutants` runs as a subprocess of the test process so that we get the most realistic view of its behavior. In some cases it is run via the `cargo` command to test that this level of indirection works properly.
+
+### `testdata` trees
+
+The primary means of testing is Rust source trees under `testdata/tree`: you can copy an existing tree and modify it to show the new behavior that you want to test.
+
+There is a general test that runs `cargo mutants --list` on each tree.
+
+A selection of test trees are available for testing different scenarios. If there is an existing suitable tree, please use it. If you need to test a situation that is not covered yet, please add a new tree.
+
+Please describe the purpose of the testdata tree in a `README.md` within the tree.
+
+To make a new tree you can copy an existing tree, but make sure to change the package name in its `Cargo.toml`.
+
+All the trees need to be mentioned in the top-level `Cargo.toml` as either included in the workspace or excluded from it. Trees that build and test successfully and that are not otherwise incompatible should be included.
+
+### Insta snapshots
+
+Many tests use Insta <https://insta.rs> to assert the expected output. Insta makes it easy to review and accept changes to expected output, either when there's a real functional change or when for example line numbers or output formatting changes.
+
+To conveniently review changed output, `cargo install cargo-insta` and then run `cargo insta test --review` etc.
+
+### Test performance
+
+CLI tests spawn a new process which can be slightly slow, especially on Windows, and especially if they actually test the mutants.
+
+So, try to keep `testdata` trees reasonably minimal for whatever they're intended to test.
+
+### Unit tests
+
+Although we primarily want to test the public interface (which is the command line), unit tests can be added in a `mod test {}` within the source tree for any behavior that is inconvenient to exercise from the command line.
+
+### Debugging code under test
```

Comment:
This is a handy section!👌

It is also good that we are tool-agnostic, because we might consider putting more info about how to do this within CLion and VSCode, but such info can be easily outdated...

Reply:
I'm happy to add anything specifically about cargo-mutants in a later PR.

Reply:
all right, please mention me in that PR, I am happy to add info about it in regards to CLion.